### DISCUSSION TITLE
Moving to Beta & Git Hooks

### DISCUSSION THREAD

**Discussion by tajmone** [tajmone/Git-Buch_EN#24] [Dev Strategies]
@SicroAtGit, we need to decide what to do with the Git hook scripts, since we're heading for the first Beta release (see #22).

- [`install-git-hooks.sh`][install-git-hooks.sh]
- [`uninstall-git-hooks.sh`][uninstall-git-hooks.sh]

I've starting working on the Rakefile in the [`alpha-dev_rake`][alpha-dev_rake] branch.

Right now, the precommit hook CDs `docs_src` and executes `./build.sh` — but the `docs_src/build.sh` script has now been deleted and replaced by the `rake preview` task, which does the same exact thing (but invoking the Asciidoctor API instead of the command line tool).

Unlike shell scripts, Rake can be invoked from any repository subfolder and will do the same job — i.e. Rake always runs from the folder where the `Rakefile` is stored.

I'm planning to make the preview document `docs_src/Git-Buch_EN.html` ignored by Git, it's only goal being for developers to peek at changes with an HTML doc that doesn't use Data URIs, so it only needs refreshing to see changed images.

Since we'll be publishing the HTML book on GHPages, will be building the HTML book in the `docs/` folder (still haven't decided the actual filename, could be `Git-Buch_EN.html`, `GitBook.html` or `Git-Book.html`, we'll discuss that). Also, this requires us to merge the book into `master`, since GHPages served from `docs/` only works on the default branch.

In any case, updating the HTML document in the `docs/` folder will be considered "publishing" the book, and will probably be assigned to `take publish` (haven't yet decided what the default task will be).

Contributors will be submitting PRs and changes to `beta-dev` branch, and the HTML book will need to be updated (published) only when merging to `master`.

IMO it would be better if contributors don't rebuild the HTML book, because it would lead to conflicts if they have to rebase, and because any differences in the Asciidoctor version employed would result in different HTML tags and CSS.

Instead, we'll be rebuilding the HTML docs before merging from `beta-dev` into `master` only — which makes sense, since the doc will only be visible online once it's on `master` branch, and for previews there's the Preview Doc (untracked).

I'm not quite sure that we should keep a Git hook that forces rebuilding the document then. I was thinking instead of having a hook that validates the code styles via EClint, but today I discovered that EClint is not working anymore on this repository (could be that the machine needs a reboot, or due to the Dia downloaded binaries which clutter the repository, although ignored).

I'll have to check if the EClint alternative tool is productions ready (last time I've checked, it still lacked some features we need, but some time has past now). Also, the new edited validation script, which only runs EClint of files which are tracked by Git, takes too long (almost five times longer) and is also buggy in various cases (deleted files, etc.). I should look into a better solution via Rake and a custom Ruby method that could filter out EClint error logs, instead of invoking Git so many times.

So, roughly the issues described above are the starting point for the switch to Rake for automating the project, and the changes required to publish the book on GHPages, which we should decide upon and sort out before moving into the Beta stage.

Even though, this Issue was supposed to be just about the Git hook scripts, it inevitably had to spread out to the general changes that are awaiting us, after all these hooks are there to serve the project even if/when it changes.

All suggestions are welcome, for most of it still mostly brainstorming.

[alpha-dev_rake]: https://github.com/tajmone/Git-Buch_EN/tree/alpha-dev_rake

[install-git-hooks.sh]: https://github.com/tajmone/Git-Buch_EN/blob/alpha-dev/install-git-hooks.sh
[uninstall-git-hooks.sh]: https://github.com/tajmone/Git-Buch_EN/blob/alpha-dev/uninstall-git-hooks.sh

**Comment by SicroAtGit** [comment]
> Unlike shell scripts, Rake can be invoked from any repository subfolder and will do the same job — i.e. Rake always runs from the folder where the `Rakefile` is stored.

That's cool.

> I'm planning to make the preview document `docs_src/Git-Buch_EN.html` ignored by Git, it's only goal being for developers to peek at changes with an HTML doc that doesn't use Data URIs, so it only needs refreshing to see changed images.

Good idea. In the last part of the sentence, I think you meant to write "so it only needs refreshing to see changed text".

> Since we'll be publishing the HTML book on GHPages, will be building the HTML book in the `docs/` folder (still haven't decided the actual filename, could be `Git-Buch_EN.html`, `GitBook.html` or `Git-Book.html`

I think `git-book.html` is good.

> Also, this requires us to merge the book into `master`, since GHPages served from `docs/` only works on the default branch.

I don't think that's true anymore. According to the GitHub documentation, the branch and directory can be selected:
[Configuring a publishing source for your GitHub Pages site](https://docs.github.com/en/pages/getting-started-with-github-pages/configuring-a-publishing-source-for-your-github-pages-site)

> `take publish` (haven't yet decided what the default task will be)

Only `publish` would be good. Default should be `preview` because this task is called the most.

> Contributors will be submitting PRs and changes to `beta-dev` branch, and the HTML book will need to be updated (published) only when merging to `master`.

Is good. Here you can maybe use Travis again to automate the HTML update.

> I'm not quite sure that we should keep a Git hook that forces rebuilding the document then. I was thinking instead of having a hook that validates the code styles via EClint

I agree.

What do you think about letting `rake preview` create the preview HTML file in the OS temp directory instead of in the repository, and then automatically open the internet browser with that preview HTML file? Then you don't need an ignore rule in `.gitignore`. And you want to open the preview HTML file in the internet browser afterwards anyway.

> but today I discovered that EClint is not working anymore on this repository (could be that the machine needs a reboot, or due to the Dia downloaded binaries which clutter the repository, although ignored).

Works here, just takes a while:
```console
$ git branch --show-current
alpha-dev_rake

$ ./validate.sh 

================================================
Validating Code Styles via EditorConfig Settings
================================================
Using:
* Node.js v16.8.0
* EClint v2.8.1.

/// Test Passed ///
```

**Reply by tajmone** [reply]
> I think you meant to write "so it only needs refreshing to see changed text".

No I meant images, because the published HTML doc uses Data URIs, i.e. images are embedded in the document via URLs and Base 64 encoding, so the doc can be downloaded as a single standalone file, images and all. But when developing the doc, you need to be able to see changed images without having to re-convert the document to HTML, hence the need to invoke Asciidoctor without the Data URIs option for the HTML Preview doc.

> I think `git-book.html` is good.

Yes, I like that more too, there's no reason why we shouldn't translate the title.

My only concern is that the (well known) _Pro Git_ book is often referred to a "The Git Book" for brevity:

https://git-scm.com/book/en/v2

so I was wondering if this could lead to confusion, or to problems finding it via search engines — googling "git book" definitely results in _Pro Git_ showing up first.

Probably there weren't many books on Git in German at that time, so a simple title like "Git Buch" could only refer to that book. In English though there are many books on Git, so a generic title like this might not play out very well unfortunately (whereas keeping part of the title in German does the trick).

Also, I didn't personally mind keeping the German title, which in any case one can intuitively guess that "buch" means "book", but then if we do so we'll have to add "English" or "EN" to avoid confusion with the German original.
Although if we include "The" in the title then there should be no confusion (_The Git Buch_ vs _Das Git Buch_).

But these consideration are mostly about the book title, not the filename (which doesn't have to match the title 100%).

> I don't think that's true anymore. According to the GitHub documentation, the branch and directory can be selected:

Right. They keep changing these settings all the time, it's hard to catch up. I'll need to actually test it though, because the screenshot in that page are hard to read so I couldn't check all the options.

So, assuming we can serve GHPages from the `docs/` folder of the Beta dev branch, do you think we should do that and then change the settings when finally reach v1.0.0?

The benefits of serving from `master` would be that in the Beta branch we could also include the final HTML doc for local preview, before merging it into `master`, which would then make it publicly visible. This would allow us to check that everything is in order, since we'd be seeing the same identical document (unlike the Preview doc, which might vary depending on the Asciidoctor version used, and its dependencies).

I don't have any problems having Beta versions in `master`, I would just like to keep that branch clutter-free as much as possible, but as the Alpha stage is over I think it should be fine even if we just merge into `master` from the beta dev branch — after all, the book is now well formed, and all the old stuff from the migration are in the Alpha branch anyhow.

> Only `publish` would be good. Default should be `preview` because this task is called the most. 

I'm not quite sure either should be default really. Bear in mind that once we start adding syntax highlighting to the book, build times will increase significantly (unless the preview is built without syntax highlighting), so it might be a feature worth making available only when one really needs to rebuild the document. Of course, it really depends on what other tasks will end up in the Rake file, which is what will ultimately decide what the default should do (usually Default includes more than one task); so right now it's mostly brainstorming.

> What do you think about letting rake preview create the preview HTML file in the OS temp directory instead of in the repository, and then automatically open the internet browser with that preview HTML file? Then you don't need an ignore rule in .gitignore. And you want to open the preview HTML file in the internet browser afterwards anyway.

Sometimes (and very often when working on docinfo files and CSS) I need to inspect the generated HTML file, either manually in the editor or via some dedicated tools, so it's more practical to have it in the same folder as the sources and ignore it. 

Also, bear in mind that the generated HTML preview file is actually a _Rake target_ (which is why if it's deleted or absent it gets created by Rake) so I think it has to be within the repository, since Rake is designed to track files within the directory tree where the `Rakefile` is located (I'm not sure if it can handle targets or sources outside that tree, I need to check this, but it would definitely be some anti-pattern).

> Works here, just takes a while:

Yes, seems to work again also on my machine, after the reboot, but it takes ages. The new script that doesn't work well with big repositories. Also, I think the problem are the SVG files, because EClint is known to have problem handling XML files (fails parsing them properly, lots of false positives). Even though there are no rules for SVG files, it's not excluding them entirely. 

E.g. I noticed that if I run the Dia downloader script, then EClint stops working. Even though the Dia bins folder is ignored by Git, I tried to prevent EClint from scanning it, but since it has no folder or file exclusion rules, all I could try was to set a match all pattern for that folder with all rules `= unset`, but that made things worst, for EClint would run forever (I'd have to brute-exit it after half hour with no result).

**Reply by SicroAtGit** [reply]
> No I meant images, because the published HTML doc uses Data URIs

But the sentence was about the preview HTML, so I was wondering why it "needs refreshing to see changed images" if it doesn't use Data URIs. Anyway, thanks for the further explanation. I already thought that you meant it that way. Good idea.

> _The Git Buch_

I like it. What else occurred to me: _The German Git Book (english)_

> So, assuming we can serve GHPages from the `docs/` folder of the Beta dev branch, do you think we should do that and then change the settings when finally reach v1.0.0?

But that requires that we continue to commit the HTML doc every time we make a change, as before. But your new idea is to stop doing that (only commit the changed AsciiDoc sources).

Maybe we can use that, then we don't need HTML files for the GH Pages anymore: [jekyll-asciidoc](https://github.com/asciidoctor/jekyll-asciidoc)

Or what do you think about using a separate orphaned branch just for the HTML doc in the beta stage? GH Pages will then generate the pages from there. Here everything in detail:

````console
$ git checkout --orphan=html-doc

$ git reset --hard

$ git commit --allow-empty -m "Initial"
````
The branch `html-doc` is then the source for GitHub Pages.

With every commit we create or merge in `beta-dev`, Travis then does this:

````console
$ git checkout beta-dev

$ rake publish

$ git checkout html-doc

$ git add docs/git-book.html

$ git commit --amend --no-edit

$ git checkout beta-dev

$ rm docs/git-book.html
````

When we finish the beta stage and release the final version, we can create a commit in the `beta-dev` branch that also includes the HTML doc and then (after we have merged the branch `beta-dev` into `master` branch) we change the source for GH Pages to the `master` branch. Finally, we can also remove the `html-doc` branch.

> > Only `publish` would be good. Default should be `preview` because this task is called the most.
> 
> (usually Default includes more than one task)

Ok, I didn't know that this is possible. Sure, then maybe it makes more sense that default (just `rake`, without parameters) stands for publish and we have the possibility to run multiple tasks in this case. With the parameters (e.g. `preview`) we can then specify more precisely which tasks should be executed.

> E.g. I noticed that if I run the Dia downloader script, then EClint stops working. Even though the Dia bins folder is ignored by Git, I tried to prevent EClint from scanning it, but since it has no folder or file exclusion rules

I saw that you use `git ls-files --exclude-standard -co` in the `validate.sh` script. For me, it does not list the contents of the `assets_src/img/dia/Dia` directory (yes, I unpacked the Windows version of Dia there).

You could additionally specify the parameter `--exclude-from=<file>` when calling `git ls-files`:
```
-X <file>, --exclude-from=<file>
Read exclude patterns from <file>; 1 per line.
```
This way you would have the exclusion functionality you are missing with EClint and can exclude more files that are not in `.gitignore`.

**Reply by tajmone** [reply]
> Maybe we can use that, then we don't need HTML files for the GH Pages anymore: jekyll-asciidoc

Unfortunately we can't use the AsciiDoc extension for Jekyll on GHPages; I don't remember if it's because it's blocked, or because you can't use custom extensions. Either way, when I last checked it wasn't usable.

> Or what do you think about using a separate orphaned branch just for the HTML doc in the beta stage? GH Pages will then generate the pages from there. 

Well, initially GHPage used to be only available from the orphan branch `gh-pages`, but now we have more options. Personally, I don't like these orphan branches, we're using this system in another documentation project, but the orphan branch tends to clutter the repository and interfere with some Git operations (sorry if I don't paste a link to an article on the problems of using orphan branches, but until the toothache is gone I have limited focusing span, but I'll dig it up, I know in which thread it came up, I just need to look for it, because we discussed the pros an cons of orphan branches before adopting them). 

Also, storing the HTML file in `docs/` allows us to see the HTML in dev branches without affecting the website. 

In our project, we had to use orphan branches for various reasons, one being the considerable size of all the documents (PDF, HTML, and Zips with code), so having them on an orphan branch avoids cluttering the repository, and we now and then squash together all the commits in the orphan branch, to regain disk space. But in general, I dislike it and avoid it if I can.

> Ok, I didn't know that this is possible. Sure, then maybe it makes more sense that default (just rake, without parameters) stands for publish and we have the possibility to run multiple tasks in this case. With the parameters (e.g. preview) we can then specify more precisely which tasks should be executed.

Yes, we have total freedom in deciding what the default tasks will be, etc. Rake follows the Make conventions here — `ALL`, `CLEAN`, `CLOBBER`, etc. being well established and expected tasks to be found in most projects.

Whether we'll add code styles validation to certain tasks will be up to us, and we can definitely make any task branch-aware, since we can invoke Git from the Rakefile, etc.

> I saw that you use `git ls-files --exclude-standard -co` in the `validate.sh` script. For me, it does not list the contents of the `assets_src/img/dia/Dia` directory (yes, I unpacked the Windows version of Dia there).

The problem is that EClint does! There are no ignore pattern that can be passed to EClint, so it will scan any subfolders (except `.git/`, which is excluded by default). Somehow this ends up being a problem under Windows (and using Bash for Windows), probably because EClint tends to clog when applying filters to long lists of files, or something similar (it's a really buggy program, and it's running under an emulated Bash for Windows, invoking Node.js and Git, so it might be too much).

> You could additionally specify the parameter `--exclude-from=<file>` when calling git ls-files:

I'll get back to this point later one (right now, fever levels prevent me from following tech details). But yes, I'm aware there's a problem and that it needs to be fixed, both with the Git options being used, but mostly because EClint fails to handle to paths with escape chars returned by Git — if a file contains an accented letter like `ò`, the whole script fails because Git will ASCII escape the `ò`, resulting in EClint complaining about a non-existent file. But this is hard to handle because different OSs might handle these differently (especially macOS), so we have no bullet proof way to handle this via the command line, but should be fixable using Ruby code in Rake instead (Ruby has many methods to handle strings, unlike Bash scripts which make such operations so painful that you start to consider founding an anti-POSIX social movement).

I'll have to postpone any code operations that require using more than 2 neurons to when the tooth extraction is over, and fever levels down to acceptable. Right now is a series of up and downs, all day long.

**Reply by SicroAtGit** [reply]
> the orphan branch tends to clutter the repository and interfere with some Git operations

Yes, I know. They have their own commits history, which has besides advantages also disadvantages. If you want to merge commits from orphaned branches into the normal commit history (e.g. into the `master` branch), you have to use `git cherry-pick` for example.

> Also, storing the HTML file in `docs/` allows us to see the HTML in dev branches without affecting the website.

I was thinking too complicated.

The simplest solution is if we create an ignore rule in `.gitignore` for `docs/*`, similar to what you mentioned and planned above. If we ran the publishing tasks with `rake` and then want to commit the HTML doc, we can use `git add -f docs/git-book.html` to force Git to include the ignored `docs/git-book.html` file in the commit.

- `docs/git-book.html` is the HTML doc to be published.

- `docs/git-book-preview.html` is the preview HTML that will never be published

If GH Pages also respects `.gitignore` and thus does not process `docs/git-book.html`, we can put `docs/git-book-preview.html` in `.gitignore` instead of `docs/*`.

> I'll have to postpone any code operations that require using more than 2 neurons to when the tooth extraction is over, and fever levels down to acceptable. Right now is a series of up and downs, all day long.

Wish you a speedy recovery.

**Comment by tajmone** [comment]
# Beta vs Final

@SicroAtGit, I think that for the Beta release we should make things simple, and just serve the GHPages website from the `docs/` folder, than when the book is finished (and we'll probably also offer a PDF) we'll switch to orphan branches.

I just wanted to make the Beta book visible so we can get contributors to participate to the polishing stage.

We'll probably need a Git hook to ensure that any commit, merge or rebase to `master` branch doesn't delete the HTML in the `docs/` folder though, because doing so (by accident) would result in the website showing a 404 error. It should be fairly simple once we've decided the HTML file name, requiring a simple file exist check.

# Book Title

The only thing that I'm still unsure about is the title translation.

I keep liking the German style title "The Git Buch", but we might have to consider that such a title could make the book harder to find via search engines.

The problem with "Git Book" is that when you search for the term the book _Pro Git_ shows up. E.g. in DuckDuckGo you get in first position:

> Git - Book
> https://git-scm.com/book
> The entire Pro Git book, written by Scott Chacon and Ben Straub and published by Apress, is available here. All content is licensed under the Creative Commons Attribution Non Commercial ...

which is why _Pro Git_ is also known as the "Git Book".

Search engines aside, I feel that keeping the German "Buch" in the title confers a strong identity to the book, but that's just my opinion.

Alternatively, we'll have to make an editorial choice and come up with some other title (which I don't really like much, since it's a translation, but it's not something unusual either).

In any case, this shouldn't be a reason to block the ongoing work, since the HTML files in the website will have to be called just `index.html` right now, and the Preview file doesn't really matter what we call it (or if we rename it later) since it's not even tracked.

I'll try to fix the remaining pending tasks from #22, and switch to Beta stage as soon as we're ready to go, so the book can finally be visible via the website and I can ask for support on the GH Community.

**Reply by tajmone** [reply]
# Done!

The simplified setup version for the Beta stage was merged to the alpha-dev branch. We can keep using this discussion for deciding how the final book will be handled in the future (orphan branches, etc.).

**Reply by SicroAtGit** [reply]
Good work!

> We'll probably need a Git hook to ensure that any commit, merge or rebase to `master` branch doesn't delete the HTML in the `docs/` folder

Good idea.

> The only thing that I'm still unsure about is the title translation.

Yes, I also need to think about it some more.

_The Git Buch_ is good. I just wonder if non-German speakers understand by _Buch_ that it's a book. Maybe it would help if we add a description metatag in the HTML and describe there that it is a book. The search engines show the text that is in the description metatag in the search results.

> The problem with "Git Book" is that when you search for the term the book _Pro Git_ shows up.

When I search with Startpage for _The Git Buch_, the result looks like this:

1. German version of ProGit
2. ProGit
3. the book we translate
4. a book that is called _Git-Buch_

However, on which position a web page appears in the search results also has to do with

- how many other pages link to it (backlink)
- what metatag keywords are set
- from which country you are searching
- what you searched for before (search history, tracking cookies)
- and so on

If I search for "The Git Buch" (in quotes) no search hit is displayed. This looks good :-)

**Reply by tajmone** [reply]
For the time being, I've opted to name the file along the "Git Book" lines, as well as the Git.io posh URL, since these two would help correct search engines indexing.

As for "The Git Buch", the repository name being `Git-Buch_EN` will ensure that the German title is also going to be present in every URL (at least in the "official" sources), so somehow both "Git Book" and "Git Buch" are already associated with the project and book right now.

I've also been thinking that in reality the title of the German book is just "Git" followed by the subtitle — i.e. _Git, Verteilte Versionskontrolle für Code und Dokumente_ (Git, Distributed version control for code and documents) — so probably "Git Buch" is a kind of alias, mostly associated to the original website domain, but in the original catalogue of the publisher the book was indeed referred to as "Git".

So we'll probably have to pick a formal title, on the one hand, and push for a short alias too ("The Git Buch"), the latter depending more on how we refer to the book in the README, links, whenever we promote it, and in our "Preface to The English Translation" (which we'll have to add to the book), etc.

The original title is definitely too short, and we'll have to come up with an alternative translation since books are catalogued mainly by the title, the subtitle being often shown only in the actual book page (I'm thinking of proper book metadata here, not search engines).

We also still have in the repo the original README in German and its translation in English ([`README_DE.md`][README_DE.md] and
[`README_EN.md`][README_EN.md]) which might contain some hints on how to pick the English title.

Since this topic still needs further thoughts, I've started a dedicated Discussion in #28, with a resume of the various points that came up so far.



[README_DE.md]: https://github.com/tajmone/Git-Buch_EN/blob/master/README_DE.md
[README_EN.md]: https://github.com/tajmone/Git-Buch_EN/blob/master/README_EN.md
